mono/utils/mono-os-mutex.h

```diff
@@ -369,26 +380,23 @@ mono_os_cond_destroy (mono_cond_t *cond)
 	/* Beauty of win32 API: do not destroy it */
 }
 
-static inline void
-mono_os_cond_wait (mono_cond_t *cond, mono_mutex_t *mutex)
-{
-	BOOL res;
-
-	res = SleepConditionVariableCS (cond, mutex, INFINITE);
-	if (G_UNLIKELY (res == 0))
-		g_error ("%s: SleepConditionVariableCS failed with error %d", __func__, GetLastError ());
-}
-
 static inline int
 mono_os_cond_timedwait (mono_cond_t *cond, mono_mutex_t *mutex, guint32 timeout_ms)
 {
-	BOOL res;
+	BOOL res = mutex->recursive ?
+		SleepConditionVariableCS (cond, &mutex->critical_section, timeout_ms) :
+		SleepConditionVariableSRW (cond, &mutex->srwlock, timeout_ms, 0);
+
+	g_assertf (res || GetLastError () == ERROR_TIMEOUT,
```

Comment:
The assert is slightly altered by sharing the code  -- previously infinite wait didn't check for error == timeout.
I can fix that in multiple ways if desired.
1. check for infinite
2. resplit the functions

Reply:
I modified the assert though. It seemed wrong.
I assume spurious wakes return true. ?

Reply:
That is not clearly documented, it only says:

> Condition variables are subject to spurious wakeups (those not associated with an explicit wake) and stolen wakeups (another thread manages to run before the woken thread). Therefore, you should recheck a predicate (typically in a while loop) after a sleep operation returns.

So I guess that means it should return true for spurious wakeups, because you need to recheck predicate meaning you can't distinguish between a "successful" and spurious wakeup.

Reply:
The documentation does not seem clear to me.
And in the examples, they do not check the return value.